File: frontend-vue/src/stores/extensions.ts
/**
 * @module stores/extensions
 * @description Event-driven extension presence registry.
 * Extensions announce themselves via SSE (extension_online / extension_offline).
 * Required/core surfaces always appear; optional ones only when running.
 *
 * Also provides API actions for the Snackbar extensions tab:
 * fetchCatalogue, toggleExtension, installExtension, repairExtension.
 */
import { ref, computed } from "vue";
import { defineStore } from "pinia";
import { SNACKBAR_BASE } from "@/api/base";

export type ExtensionKind = "core" | "surface" | "service" | "plugin";
export type ExtensionStatus = "unknown" | "available" | "installed" | "running";

export interface ExtensionManifest {
  id: string;
  name: string;
  kind: ExtensionKind;
  required: boolean; // core surfaces always shown
  route?: string;
  icon: string;
  description?: string;
  status_endpoint?: string;
  install_url?: string;
  activation_required: boolean;
  version?: string;
}

interface ExtensionEntry {
  manifest: ExtensionManifest;
  status: ExtensionStatus;
  lastSeen?: number; // epoch ms
}

// ─── Built-in manifest registry ─────────────────────────────────

const BUILTIN_MANIFESTS: ExtensionManifest[] = [
  // Core surfaces — always visible
  {
    id: "dashboard",
    name: "Dashboard",
    kind: "core",
    required: true,
    icon: "home",
    route: "/",
    activation_required: false,
  },
  {
    id: "workflow",
    name: "Workflow",
    kind: "core",
    required: true,
    icon: "flag",
    route: "/workflow",
    activation_required: false,
  },
  {
    id: "intelligence",
    name: "Intelligence",
    kind: "core",
    required: true,
    icon: "lightbulb",
    route: "/intelligence",
    activation_required: false,
  },
  {
    id: "system",
    name: "System",
    kind: "core",
    required: true,
    icon: "settings",
    route: "/system",
    activation_required: false,
  },
  {
    id: "snackbar",
    name: "Snackbar",
    kind: "core",
    required: true,
    icon: "storefront",
    route: "/snackbar",
    activation_required: false,
  },
  // Optional surfaces — supply on demand
  {
    id: "browserui",
    name: "Browser",
    kind: "surface",
    required: false,
    icon: "language",
    route: "/browserui",
    activation_required: false,
    description: "Web research and bookmark stacks",
  },
  {
    id: "ucode",
    name: "uCode",
    kind: "surface",
    required: false,
    icon: "grid_on",
    route: "/ucode",
    activation_required: false,
    description: "Grid engine, Teletext, terminal",
  },
  {
    id: "groovebox",
    name: "Groovebox",
    kind: "surface",
    required: false,
    icon: "music_note",
    route: "/groovebox",
    activation_required: false,
    description: "Music and audio tools",
  },
  {
    id: "sonic",
    name: "Sonic Screwdriver",
    kind: "surface",
    required: false,
    icon: "build_circle",
    route: "/sonic",
    activation_required: false,
    description: "System diagnostics and repair",
  },
  {
    id: "documentation",
    name: "Documentation",
    kind: "surface",
    required: true,
    icon: "menu_book",
    route: "/documentation",
    activation_required: false,
    description: "Built-in documentation viewer",
  },
  // Extensions that require installation + activation
  {
    id: "udev",
    name: "Developer",
    kind: "surface",
    required: false,
    icon: "code",
    route: "/developer",
    activation_required: true,
    description: "Full developer lane — requires uDev",
    install_url: "https://github.com/fredporter/uDev",
  },
  {
    id: "snack-shack",
    name: "Snack Shack",
    kind: "plugin",
    required: false,
    icon: "storefront",
    route: "/snackbar?tab=snacks",
    activation_required: false,
    description: "Browse and install AI model packages",
  },
  {
    id: "plugin-store",
    name: "Plugin Store",
    kind: "plugin",
    required: false,
    icon: "extension",
    route: "/intelligence?tab=plugins",
    activation_required: false,
    description: "Discover and install uDOS extensions",
    install_url: "#",
  },
  // External udos-* extension manifest entries
  {
    id: "udos-budget",
    name: "uDos Budget",
    kind: "plugin",
    required: false,
    icon: "savings",
    activation_required: false,
    description: "Budget policy & status plugin",
  },
  {
    id: "udos-identity",
    name: "uDos Identity",
    kind: "plugin",
    required: false,
    icon: "fingerprint",
    activation_required: false,
    description: "Identity profile & session plugin",
  },
  {
    id: "udos-google",
    name: "Google Bridge",
    kind: "plugin",
    required: false,
    icon: "cloud",
    activation_required: false,
    description: "Google OAuth, Gemini/Gems, Drive mirror",
  },
  {
    id: "udos-dreamscape",
    name: "Dreamscape",
    kind: "plugin",
    required: false,
    icon: "psychology",
    activation_required: false,
    description: "Mission scaffolding & daily briefing",
  },
  {
    id: "udos-publishing",
    name: "Publishing",
    kind: "plugin",
    required: false,
    icon: "publish",
    activation_required: false,
    description: "Cloud mirror for udo.guide/udo.place",
  },
  {
    id: "udos-vaults",
    name: "Vault Topology",
    kind: "plugin",
    required: false,
    icon: "folder_special",
    activation_required: false,
    description: "Vault topology & AppFlowy bridge",
  },
  {
    id: "udos-agents",
    name: "uDos Agents",
    kind: "plugin",
    required: false,
    icon: "smart_toy",
    activation_required: false,
    description: "Specialized agent scaffolding",
  },
  {
    id: "homenest",
    name: "HomeNest",
    kind: "plugin",
    required: false,
    icon: "home",
    activation_required: false,
    description: "Home stream server — Jellyfin + Home Assistant bridge",
  },
];

// ─── Store ───────────────────────────────────────────────────────

export const useExtensionStore = defineStore("extensions", () => {
  const localEnabled = ref<Record<string, boolean>>({});
  const entries = ref<Map<string, ExtensionEntry>>(new Map());

  // Initialise all known manifests as "unknown"
  for (const manifest of BUILTIN_MANIFESTS) {
    entries.value.set(manifest.id, {
      manifest,
      status: manifest.required ? "running" : "unknown",
    });
  }

  function applyLocalEnabledOverrides() {
    for (const [id, enabled] of Object.entries(localEnabled.value || {})) {
      const entry = entries.value.get(id);
      if (!entry || entry.manifest.required) continue;
      entry.status = enabled ? "running" : "installed";
    }
  }

  applyLocalEnabledOverrides();

  // ─── Computed ───────────────────────────────────────────────

  const all = computed(() => [...entries.value.values()]);

  const running = computed(() =>
    all.value.filter((e) => e.status === "running"),
  );

  /** Surfaces to show in Dashboard + nav: required OR running */
  const visibleSurfaces = computed(() =>
    all.value
      .filter(
        (e) => e.manifest.kind === "surface" || e.manifest.kind === "core",
      )
      .filter((e) => e.manifest.required || e.status === "running"),
  );

  /** All known extensions for the plugin catalogue */
  const catalogue = computed(() =>
    all.value.filter((e) => !e.manifest.required),
  );

  const isRunning = (id: string) => entries.value.get(id)?.status === "running";

  const getStatus = (id: string): ExtensionStatus =>
    entries.value.get(id)?.status ?? "unknown";

  // ─── Mutations ───────────────────────────────────────────────

  function markRunning(id: string, version?: string) {
    const entry = entries.value.get(id);
    if (entry) {
      entry.status = "running";
      entry.lastSeen = Date.now();
      if (version) entry.manifest.version = version;
    } else {
      // Unknown extension announcing itself — register dynamically
      entries.value.set(id, {
        manifest: {
          id,
          name: id,
          kind: "plugin",
          required: false,
          icon: "extension",
          activation_required: false,
        },
        status: "running",
        lastSeen: Date.now(),
      });
    }
  }

  function markOffline(id: string) {
    const entry = entries.value.get(id);
    if (entry && !entry.manifest.required) {
      entry.status = "installed"; // was running, now stopped
    }
  }

  function markInstalled(id: string) {
    const entry = entries.value.get(id);
    if (entry) entry.status = "installed";
  }

  /** Prune entries not seen in last 60s (stale running) */
  function pruneStale() {
    const cutoff = Date.now() - 60_000;
    for (const [id, entry] of entries.value) {
      if (
        entry.status === "running" &&
        entry.lastSeen &&
        entry.lastSeen < cutoff &&
        !entry.manifest.required
      ) {
        entry.status = "installed";
      }
    }
  }

  // ─── API-driven catalogue & actions (Snackbar Extensions tab) ──

  /** Runtime catalogue enriched with backend probe data */
  const runtimeCatalogue = ref<any[]>([]);
  /** Loading states per action */
  const loading = ref<Record<string, string>>({}); // id -> action
  const actionMessage = ref<string>("");

  async function fetchCatalogue() {
    try {
      const res = await fetch(`${SNACKBAR_BASE}/api/extensions/catalogue`, {
        signal: AbortSignal.timeout(8000),
      });
      if (!res.ok) return;
      const data = await res.json();
      runtimeCatalogue.value = data.extensions ?? [];

      // Sync status back into known entries
      for (const ext of runtimeCatalogue.value) {
        if (typeof ext?.id === "string") {
          localEnabled.value = {
            ...localEnabled.value,
            [ext.id]: Boolean(ext.enabled),
          };
        }

        const existing = entries.value.get(ext.id);
        if (existing) {
          if (ext.status === "running" || ext.enabled) {
            existing.status = "running";
          } else if (ext.is_installed) {
            existing.status = "installed";
          } else {
            existing.status = "available";
          }
          if (ext.version) existing.manifest.version = ext.version;
        }
      }

      applyLocalEnabledOverrides();
    } catch {
      // Backend unreachable — stick with built-in status
      applyLocalEnabledOverrides();
    }
  }

  async function toggleExtension(id: string, enabled: boolean) {
    loading.value = { ...loading.value, [id]: "toggling" };

    localEnabled.value = {
      ...localEnabled.value,
      [id]: enabled,
    };

    // Optimistic update — set local status immediately
    const existing = entries.value.get(id);
    if (existing) {
      existing.status = enabled ? "running" : "installed";
    }

    try {
      const res = await fetch(`${SNACKBAR_BASE}/api/extensions/${id}/toggle`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ enabled }),
        signal: AbortSignal.timeout(5000),
      });
      const data = await res.json();
      actionMessage.value = data.message ?? "";
    } catch {
      // Backend unreachable — optimistic state is already set
      actionMessage.value = "";
    } finally {
      loading.value = { ...loading.value, [id]: "" };
    }
  }

  async function installExtension(id: string) {
    loading.value = { ...loading.value, [id]: "installing" };
    try {
      const res = await fetch(`${SNACKBAR_BASE}/api/extensions/${id}/install`, {
        method: "POST",
        signal: AbortSignal.timeout(120_000), // 2min timeout for clone+install
      });
      const data = await res.json();
      actionMessage.value = data.message ?? "";
      if (data.success) {
        const existing = entries.value.get(id);
        if (existing) {
          existing.status = "running";
        }
      }
    } catch (e: any) {
      actionMessage.value = `Install failed: ${e.message}`;
    } finally {
      loading.value = { ...loading.value, [id]: "" };
      // Refresh catalogue after install
      await fetchCatalogue();
    }
  }

  async function repairExtension(id: string) {
    loading.value = { ...loading.value, [id]: "repairing" };
    try {
      const res = await fetch(`${SNACKBAR_BASE}/api/extensions/${id}/repair`, {
        method: "POST",
        signal: AbortSignal.timeout(120_000),
      });
      const data = await res.json();
      actionMessage.value = data.message ?? "";
      if (data.success) {
        const existing = entries.value.get(id);
        if (existing) {
          existing.status = "running";
        }
      }
    } catch (e: any) {
      actionMessage.value = `Repair failed: ${e.message}`;
    } finally {
      loading.value = { ...loading.value, [id]: "" };
      await fetchCatalogue();
    }
  }

  return {
    localEnabled,
    entries,
    all,
    running,
    visibleSurfaces,
    catalogue,
    isRunning,
    getStatus,
    markRunning,
    markOffline,
    markInstalled,
    pruneStale,
    // API-driven
    runtimeCatalogue,
    loading,
    actionMessage,
    fetchCatalogue,
    toggleExtension,
    installExtension,
    repairExtension,
  };
}, {
  persist: {
    key: "ucore.extensions",
    paths: ["localEnabled"],
  },
});

File: backend/app/snackbar/modules/extensions.py
"""Snackbar Extensions module — REST API for managing extensions.

Provides:
  GET  /api/extensions/catalogue   — full extension catalogue with status
  POST /api/extensions/{id}/toggle — enable/disable an extension
  POST /api/extensions/{id}/install — install (clone + setup) an extension
  POST /api/extensions/{id}/repair  — repair (re-clone, reinstall, fix deps)
"""
from __future__ import annotations

import json
import logging
import subprocess
from pathlib import Path

from aiohttp import web

from app.core.settings import settings

log = logging.getLogger("ucore.snackbar.extensions")

STATE_FILE = settings.data_dir / "extension_state.json"

# ── Known external extensions we expect in ~/Code/ ───────────────────

KNOWN_EXTERNAL: list[dict] = [
    {
        "id": "udos-budget",
        "name": "uDos Budget",
        "kind": "plugin",
        "icon": "savings",
        "description": "Budget policy & status plugin",
        "repo": "udos-budget",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "udos-identity",
        "name": "uDos Identity",
        "kind": "plugin",
        "icon": "fingerprint",
        "description": "Identity profile & session plugin",
        "repo": "udos-identity",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "udos-google",
        "name": "Google Bridge",
        "kind": "plugin",
        "icon": "cloud",
        "description": "Google OAuth, Gemini/Gems, Drive mirror",
        "repo": "udos-google",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "udos-dreamscape",
        "name": "Dreamscape",
        "kind": "plugin",
        "icon": "psychology",
        "description": "Mission scaffolding & daily briefing",
        "repo": "udos-dreamscape",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "udos-publishing",
        "name": "Publishing",
        "kind": "plugin",
        "icon": "publish",
        "description": "Cloud mirror for udo.guide/udo.place",
        "repo": "udos-publishing",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "udos-vaults",
        "name": "Vault Topology",
        "kind": "plugin",
        "icon": "folder_special",
        "description": "Vault topology & AppFlowy bridge",
        "repo": "udos-vaults",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "udos-agents",
        "name": "uDos Agents",
        "kind": "plugin",
        "icon": "smart_toy",
        "description": "Specialized agent scaffolding",
        "repo": "udos-agents",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
    {
        "id": "homenest",
        "name": "HomeNest",
        "kind": "plugin",
        "icon": "home",
        "description": "Home stream server — Jellyfin + Home Assistant bridge",
        "repo": "HomeNest",
        "manifest_path": "ucore-extension.json",
        "deps": [],
    },
]


# ── State persistence ────────────────────────────────────────────────


def _load_state() -> dict[str, dict]:
    """Load persisted extension state from disk."""
    if not STATE_FILE.exists():
        return {}
    try:
        return json.loads(STATE_FILE.read_text(encoding="utf-8"))
    except Exception:
        return {}


def _save_state(state: dict[str, dict]) -> None:
    """Persist extension state to disk."""
    STATE_FILE.parent.mkdir(parents=True, exist_ok=True)
    STATE_FILE.write_text(
        json.dumps(state, indent=2, default=str),
        encoding="utf-8",
    )


# ── Filesystem probes ────────────────────────────────────────────────


def _repo_exists(repo_name: str) -> bool:
    """Check if a repo directory exists under ~/Code/."""
    return (settings.udos_root / repo_name).is_dir()


def _check_manifest(repo_name: str, manifest_path: str) -> bool:
    """Check if the extension manifest file exists in the repo."""
    return (settings.udos_root / repo_name / manifest_path).is_file()


def _get_pyproject(repo_name: str) -> dict | None:
    """Read pyproject.toml from a repo (best-effort version extraction)."""
    pp = settings.udos_root / repo_name / "pyproject.toml"
    if not pp.is_file():
        return None
    try:
        text = pp.read_text(encoding="utf-8")
        # Quick regex-free parse for version
        version = None
        name = None
        for line in text.splitlines():
            stripped = line.strip()
            if stripped.startswith("version"):
                version = (
                    stripped.split("=", 1)[1]
                    .strip()
                    .strip('"')
                    .strip("'")
                )
            if stripped.startswith("name"):
                name = stripped.split("=", 1)[1].strip().strip('"').strip("'")
        return {"name": name, "version": version}
    except Exception:
        return None


def _probe_extension(ext: dict) -> dict:
    """Probe filesystem for extension presence and status."""
    result = dict(ext)
    repo = ext.get("repo", ext["id"])
    result["repo_exists"] = _repo_exists(repo)
    manifest_file = ext.get("manifest_path", "ucore-extension.json")
    result["manifest_exists"] = _check_manifest(repo, manifest_file)
    result["is_installed"] = (
        result["repo_exists"] and result["manifest_exists"]
    )

    # Try to get version from pyproject.toml
    pp = _get_pyproject(repo)
    if pp and pp.get("version"):
        result["version"] = pp["version"]

    return result


# ── Action helpers ───────────────────────────────────────────────────


def _toggle_extension(ext_id: str, enabled: bool) -> tuple[bool, str]:
    """Toggle an extension on or off in persisted state."""
    state = _load_state()
    entry = state.get(ext_id, {})
    entry["enabled"] = enabled
    entry["updated_at"] = str(subprocess.check_output(["date", "-Iseconds"]).decode().strip()) if False else ""
    state[ext_id] = entry
    _save_state(state)
    return True, f"Extension '{ext_id}' {'enabled' if enabled else 'disabled'}"


def _install_extension(ext: dict) -> tuple[bool, str]:
    """Install an extension — clone repo if missing."""
    repo_name = ext.get("repo", ext["id"])
    repo_dir = settings.udos_root / repo_name

    if repo_dir.is_dir():
        # Already exists — run install
        return _run_install_commands(repo_dir, ext)

    # Clone from GitHub
    log.info("Installing extension %s — cloning repo %s", ext["id"], repo_name)
    result = subprocess.run(
        [
            "git", "clone",
            f"https://github.com/fredporter/{repo_name}.git",
            str(repo_dir),
        ],
        capture_output=True, text=True, timeout=60,
        cwd=str(settings.udos_root),
    )
    if result.returncode != 0:
        log.error("Clone failed for %s: %s", repo_name, result.stderr)
        # Try SSH fallback
        result2 = subprocess.run(
            [
                "git", "clone",
                f"git@github.com:fredporter/{repo_name}.git",
                str(repo_dir),
            ],
            capture_output=True, text=True, timeout=60,
            cwd=str(settings.udos_root),
        )
        if result2.returncode != 0:
            return False, f"Failed to clone {repo_name}: {result2.stderr.strip()}"

    return _run_install_commands(repo_dir, ext)


def _run_install_commands(repo_dir: Path, ext: dict) -> tuple[bool, str]:
    """Run pip install or npm install in the extension repo."""
    messages: list[str] = []

    # Python install
    if (repo_dir / "pyproject.toml").is_file() or (repo_dir / "setup.py").is_file() or (repo_dir / "requirements.txt").is_file():
        log.info("Running pip install in %s", repo_dir)
        result = subprocess.run(
            ["pip", "install", "-e", str(repo_dir)],
            capture_output=True, text=True, timeout=120,
        )
        if result.returncode == 0:
            messages.append("pip install ✓")
        else:
            messages.append(f"pip install error: {result.stderr.strip()[-200:]}")

    # Node install
    if (repo_dir / "package.json").is_file():
        log.info("Running npm install in %s", repo_dir)
        result = subprocess.run(
            ["npm", "install"],
            capture_output=True, text=True, timeout=120,
            cwd=str(repo_dir),
        )
        if result.returncode == 0:
            messages.append("npm install ✓")
        else:
            messages.append(f"npm install error: {result.stderr.strip()[-200:]}")

    # Mark as enabled in state
    _toggle_extension(ext["id"], True)

    if messages:
        return True, "; ".join(messages)
    return True, "Install complete (no package manager files found)"


def _repair_extension(ext: dict) -> tuple[bool, str]:
    """Repair — re-clone repo + reinstall deps."""
    repo_name = ext.get("repo", ext["id"])
    repo_dir = settings.udos_root / repo_name

    # Re-clone: remove and re-clone
    if repo_dir.is_dir():
        log.info("Repairing %s — removing existing repo", repo_dir)
        subprocess.run(["rm", "-rf", str(repo_dir)], timeout=30)

    return _install_extension(ext)


# ── Route handlers ───────────────────────────────────────────────────


async def handle_extensions_catalogue(request: web.Request) -> web.Response:
    """GET /api/extensions/catalogue — full extension catalogue with filesystem status."""
    state = _load_state()
    catalogue = []
    for ext in KNOWN_EXTERNAL:
        entry = _probe_extension(ext)
        ext_state = state.get(entry["id"], {})
        entry["enabled"] = ext_state.get("enabled", entry["is_installed"])
        entry["status"] = _derive_status(entry)
        catalogue.append(entry)

    return web.json_response({
        "extensions": catalogue,
        "total": len(catalogue),
    })


def _derive_status(entry: dict) -> str:
    """Derive a human-readable status from probe results."""
    if entry.get("status"):
        return entry["status"]
    if entry.get("enabled") and entry.get("is_installed"):
        return "running"
    if entry.get("is_installed"):
        return "installed"
    return "available"


async def handle_extension_toggle(request: web.Request) -> web.Response:
    """POST /api/extensions/{id}/toggle — toggle extension on/off."""
    ext_id = request.match_info.get("id", "").strip()
    if not ext_id:
        return web.json_response({"error": "Extension id required"}, status=400)

    try:
        body = await request.json()
    except Exception:
        body = {}
    enabled = bool(body.get("enabled", body.get("enable", True)))

    success, message = _toggle_extension(ext_id, enabled)
    payload = {
        "id": ext_id,
        "enabled": enabled,
        "success": success,
        "message": message,
    }
    return web.json_response(payload)


async def handle_extension_install(request: web.Request) -> web.Response:
    """POST /api/extensions/{id}/install — install an extension."""
    ext_id = request.match_info.get("id", "").strip()
    if not ext_id:
        return web.json_response({"error": "Extension id required"}, status=400)

    ext = next((e for e in KNOWN_EXTERNAL if e["id"] == ext_id), None)
    if not ext:
        return web.json_response({"error": f"Unknown extension: {ext_id}"}, status=404)

    success, message = _install_extension(ext)
    payload = {
        "id": ext_id,
        "installed": success,
        "success": success,
        "message": message,
    }
    return web.json_response(payload)


async def handle_extension_repair(request: web.Request) -> web.Response:
    """POST /api/extensions/{id}/repair — repair an extension."""
    ext_id = request.match_info.get("id", "").strip()
    if not ext_id:
        return web.json_response({"error": "Extension id required"}, status=400)

    ext = next((e for e in KNOWN_EXTERNAL if e["id"] == ext_id), None)
    if not ext:
        return web.json_response({"error": f"Unknown extension: {ext_id}"}, status=404)

    success, message = _repair_extension(ext)
    payload = {
        "id": ext_id,
        "repaired": success,
        "success": success,
        "message": message,
    }
    return web.json_response(payload)


def register(app: web.Application) -> None:
    """Register extension management routes on the snackbar app."""
    app.router.add_get("/api/extensions/catalogue", handle_extensions_catalogue)
    app.router.add_post("/api/extensions/{id}/toggle", handle_extension_toggle)
    app.router.add_post("/api/extensions/{id}/install", handle_extension_install)
    app.router.add_post("/api/extensions/{id}/repair", handle_extension_repair)
    log.info("Snackbar extensions module registered (4 routes)")
